src/Common/FileCache.cpp

```diff
@@ -751,10 +926,12 @@ void LRUFileCache::removeIfReleasable(bool remove_persistent_files)
 
     std::lock_guard cache_lock(mutex);
 
-    std::vector<FileSegment *> to_remove;
```

Comment:
Let's have raw pointers instead of shared. The point is -- `file_segment->complete` method can rely on `file_segment.use_count()` which will not work correctly if here we use a shared pointer.
(looks like before it was also somewhat broken)

Reply:
There is still shared pointer

Reply:
```
    for (auto & file_segment : to_remove)
    {
        std::lock_guard segment_lock(file_segment->mutex);
        file_segment->detach(cache_lock, segment_lock);
        remove(file_segment->key(), file_segment->offset(), cache_lock, segment_lock);
    }
```
Maybe this reference is only temporary? If the shared pointer is not set, the lock of the FileSegment needs to be acquired when calling remove() to delete it, and then after deleting the FileSegment in remove() will cause an exception to be triggered when the lock is released (It will be automatically released, and its internal lock will also be released, but we hold its lock before executing remove). Code elsewhere in the deletion of a FileSegment will build a temporary shared pointer to it.

Reply:
Right, forgot about it.